# **MscPoject** - UKBB

## Table of Contents

- [**MscPoject** - UKBB](#mscpoject---ukbb)
  - [Table of Contents](#table-of-contents)
  - [**Brief Description**-***Project code***](#brief-description-project-code)
  - [**Languages**](#languages)
  - [**Installation**](#installation)
  - [**Project structure and Usage**](#project-structure-and-usage)
    - [**(1) Repo Structure introduction**](#1-repo-structure-introduction)
  - [**Author and Contact**](#author-and-contact)

## **Brief Description**-***Project code***

1. Unix and shell scripting for bioinformatics analysis. Python as a tool to control the pipeline and R for a statistical tool and visualization.
2. LaTex to do some of the scientific writing.
3. You can find the data used in this study in ./Data_Available
4. You can find the code used in this study for Microbiome analysis in ./UKBB_Microbiome, and RAD-seq analysis in 

| Rep Name |Description | 
| ------ | ------ | 
|Data_Available   | Data used in this study| 
| UKBB_Microbiomes | Code of Microbiome analysis| 
| UKBB_RADseq  | Code of RAD-seq | 


## **Languages**
```
Shell,Latex,Python,R
```
## **Installation**
```
git clone https://github.com/nedchen2/CMEECourseWork.git
```

## **Project structure and Usage**

### **(1) Repo Structure introduction**

Each directory contain directories called `code`, `data`

`code`: directory to put the scripts (`e.g. *.sh`)

`Data`: directory to put some of the input file (`e.g. *.fasta`)


## **Author and Contact**

**Congjia Chen**

Congjia.Chen21@imperial.ac.uk

